File: C++/BucketSort.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
#include <cmath>
#define vec vector<int>
using namespace std;

void PrintVec(vector<int> arr)
{
    for (auto &i : arr)
        cout << i << '\t';
    cout << endl;
}
int inputNumber()
{
    int num;
    cin >> num;
    if (cin.fail())
    {
        cin.clear();
        cin.ignore(numeric_limits<streamsize>::max(), '\n');
        throw runtime_error("INPUT MUST BE A NUMBER");
    }
    return num;
}

void BucketSort(vec &arr)
{
    int size = arr.size();
    if (size <= 0)
        return;
    int max_val = *max_element(arr.begin(), arr.end());
    int min_val = *min_element(arr.begin(), arr.end());

    int range = max_val - min_val + 1;
    vector<vector<int>> buckets(range);

    for (int i = 0; i < size; i++)
    {
        int index = arr[i] - min_val;
        buckets[index].push_back(arr[i]);
    }

    for (int i = 0; i < range; i++)
        sort(buckets[i].begin(), buckets[i].end());

    int index = 0;
    for (int i = 0; i < range; i++)
    {
        for (int j = 0; j < buckets[i].size(); j++)
        {
            arr[index++] = buckets[i][j];
        }
    }
}

int main()
{
    system("cls");
    int size, el;
    try
    {
        cout << "Enter size of the array: ";
        size = inputNumber();
        if (size <= 0)
            throw runtime_error("Size of list can not be negative or zero");
        vec arr;
        for (int i = 0; i < size; i++)
            el = inputNumber(), arr.push_back(el);
        cout << "INPUT LIST:" << endl;
        PrintVec(arr);
        BucketSort(arr);
        cout << "SORTED LIST:" << endl;
        PrintVec(arr);
    }
    catch (runtime_error e)
    {
        cout << "Error : " << e.what() << endl;
    }
    return 0;
}

File: Python/BucketSort.py
def InsertionSort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key


def BucketSort(arr):
    if len(arr) == 0:
        return arr

    min_val, max_val = min(arr), max(arr)
    num_buckets = max(max_val - min_val, 1)
    bucket_range = (max_val - min_val) / num_buckets

    buckets = [[] for _ in range(num_buckets)]

    for num in arr:
        index = int((num - min_val) / bucket_range)
        if index == num_buckets:
            index -= 1
        buckets[index].append(num)

    sorted_arr = []
    for bucket in buckets:
        InsertionSort(bucket)
        sorted_arr.extend(bucket)

    return sorted_arr


try:
    arr = []
    size = int(input("Enter the size of the list: "))
    for _ in range(size):
        while True:
            try:
                el = int(input("Enter a number: "))
                arr.append(el)
                break
            except ValueError:
                print("Invalid input. Please enter a valid number.")

    print("INPUT LIST:", arr)
    sorted_arr = BucketSort(arr)
    print("SORTED LIST:", sorted_arr)
except KeyboardInterrupt:
    print("Program interrupted by the user.")
except Exception as e:
    print(f"An error occurred: {e}")
